src/events/dns.js

```diff
@@ -0,0 +1,239 @@
+                resolve();
+                if (callback) callback(err, hostname, service);
+            };
+        });
+        clientRequest = original.apply(this, [address, port, patchedCallback]);
+        tracer.addEvent(dnsEvent, responsePromise);
+    } catch (err) {
+        tracer.addException(err);
+    }
+    if (!clientRequest) {
+        clientRequest = original.apply(this, buildParams(address, port, callback));
+    }
+    return clientRequest;
+};
+
+const wrapDnsReverseFunction = original => (ip, callback) => {
+    let patchedCallback;
+    let clientRequest;
+    try {
+        const { dnsEvent, startTime } = initialDnsEvent(original.name);
+        eventInterface.addToMetadata(dnsEvent, { ip });
+        const responsePromise = new Promise((resolve) => {
+            patchedCallback = (err, hostnames) => {
+                endDurationAndAddCallbackMetadata(dnsEvent, startTime, err, { hostnames });
+                resolve();
+                if (callback) callback(err, hostnames);
+            };
+        });
+        clientRequest = original.apply(this, [ip, patchedCallback]);
+        tracer.addEvent(dnsEvent, responsePromise);
+    } catch (err) {
+        tracer.addException(err);
+    }
+    if (!clientRequest) {
+        clientRequest = original.apply(this, [ip, callback]);
+    }
+    return clientRequest;
+};
+
+const wrapDnsLookupFunction = original => (arg1, arg2, arg3) => {
+    let patchedCallback;
+    let clientRequest;
+    const { hostname, options, callback } = getArguments(arg1, arg2, arg3);
+    if (isBlacklistURL(hostname)) {
+        utils.debugLog(`filtered blacklist hostname ${hostname}`);
+        return original.apply(this, buildParams(arg1, arg2, arg3));
+    }
+    try {
+        const { dnsEvent, startTime } = initialDnsEvent(original.name);
+        eventInterface.addToMetadata(dnsEvent, { hostname });
+        if (options) eventInterface.addToMetadata(dnsEvent, { options });
+        const responsePromise = new Promise((resolve) => {
+            patchedCallback = (err, address, family) => {
+                endDurationAndAddCallbackMetadata(dnsEvent, startTime, err, { address, family });
+                resolve();
+                if (callback) callback(err, address, family);
+            };
+        });
+        clientRequest = original.apply(this, buildParams(hostname, options, patchedCallback));
+        tracer.addEvent(dnsEvent, responsePromise);
+    } catch (err) {
+        tracer.addException(err);
+    }
+    if (!clientRequest) {
+        clientRequest = original.apply(this, buildParams(arg1, arg2, arg3));
+    }
+    return clientRequest;
+};
+
+module.exports = {
+    /**
+     * Initializes the dns tracer
+     */
+    init() {
+        // eslint-disable-next-line no-debugger
+        debugger;
+        if ((process.env.DNS_INSTRUMENTATION || '').toUpperCase() === 'TRUE') {
+            Object.keys(rrtypesMethods).forEach((functionToTrace) => {
```

Comment:
did you look at the implementation? are you 100% sure they are not calling the same function underneath and we have to patch them all? (an argument of "they might but patching them all is easy and is less prone to break because of implementation changes" might be valid here, just making sure this is your thought process)

Reply:
All those functions (rrtypesMethods keys) calls to their own function and i catch them all in dns.Resolve, I'm checking now if dns library contains each function and just then wraps it with shimmer.